orttraining/tools/scripts/pipeline_model_split.py

```diff
@@ -0,0 +1,376 @@
+
+    return new_send, new_receive
+
+def find_all_input_nodes(model, node):
+    nodes = []
+    inputs = []
+
+    if node:
+        for inputId in node.input:
+            for node in model.graph.node:
+                for output in node.output:
+                    if output == inputId:
+                        nodes.append(node)
+            for input in model.graph.input:
+                if input.name == inputId:
+                    inputs.append(input)
+    return nodes, inputs
+
+def find_all_output_nodes(model, node):
+    nodes = []
+    outputs = []
+    if node:
+        for outputId in node.output:
+            for node in model.graph.node:
+                for input in node.input:
+                    if input == outputId:
+                        nodes.append(node)
+            for output in model.graph.output:
+                if output.name == outputId:
+                    outputs.append(output)
+    return nodes, outputs
+
+def find_all_output_nodes_by_edge(model, arg):
+    result = []
+    for node in model.graph.node:
+        for input in node.input:
+            if input == arg:
+                result.append(node)
+    return result
+
+# Insert identity nodes to separate same output edge which feeds into different sub-graph.
+def add_identity(model, cuttingEdge, newEdgeIdName):
+    output_nodes = None
+    edgeId = cuttingEdge.edgeId
+    for node in model.graph.node:
+        if len(node.output) >=1 and node.output[0] == edgeId:
+            output_nodes = find_all_output_nodes_by_edge(model, node.output[0])
+            break
+
+    assert output_nodes, "no output node"
+
+    new_identity = model.graph.node.add()
+    new_identity.op_type = 'Identity'
+
+    new_identity.input.extend([edgeId])
+    new_identity.output.extend([newEdgeIdName])
+
+    for i in range(len(output_nodes)):
+        if output_nodes[i].output[0] in cuttingEdge.consumingNodes:
+            for j in range(len(output_nodes[i].input)):
+                if output_nodes[i].input[j] == edgeId:
+                    output_nodes[i].input[j] = newEdgeIdName
+
+    return newEdgeIdName
+
+def find_all_connected_nodes(model, node):
+    nodes0, inputs = find_all_input_nodes(model, node)
+    nodes1, outputs = find_all_output_nodes(model, node)
+
+    connected_nodes = nodes0 + nodes1
+    return connected_nodes, inputs, outputs
+
+def get_node_index(model, node):
+    for i in range(len(model.graph.node)):
+        if model.graph.node[i] == node:
+            return i
+    return None
+
+def get_input_index(model, input):
+    for i in range(len(model.graph.input)):
+        if model.graph.input[i] == input:
+            return i
+    return None
+
+def get_output_index(model, output):
+    for i in range(len(model.graph.output)):
+        if model.graph.output[i] == output:
+            return i
+    return None
+
+# traverse the graph, group connected nodes and generate subgraph
+def generate_subgraph(model, start0, start1):
```

Comment:
>start0, start1 [](start = 29, length = 14)

could there be more than two starts?

Reply:
yep. Now multiple cut is supported that we have multiple starts.